keps/sig-scheduling/3633-matchlabelselectors-to-podaffinity/README.md

```diff
@@ -263,8 +266,36 @@ metadata:
             values:
             - database
         topologyKey: topology.kubernetes.io/zone
-        matchlabelKeys: # ADDED
-        - pod-template-hash
+        matchLabelSelectors: # ADDED
+        - Key: pod-template-hash
+          operator: In
+```
+
+#### Story 2
+
+Let's say all Pods on each tenant get `tenant` label via a controller or a manifest management tool like Helm. 
+And, the cluster admin now wants to achieve exclusive 1:1 tenant to domain placement.
+
+By applying the following affinity globally using a mutating webhook, the cluster admin can ensure that the Pods from the same tenant will land on the same domain exclusively, meaning Pods from other `tenants` won't land on the same domain. 
+
+```yaml
+affinity:
+  podAffinity:      # ensures the pods of this tenant land on the same node pool
+    requiredDuringSchedulingIgnoredDuringExecution:
+      - matchLabelSelectors:
+          - Key: tenant
+            operator: In
```

Comment:
About the previous discussion of adding a new enum to label selector, I don't think we should do that, as existing clients are not prepared to handle that. Additionally, that new enum would make label selectors contextual, which would mean all code paths taking label selectors and returning matchers would then need to take label selectors AND a set of the labels of the containing object... that's a super invasive change I don't think we would ever successfully ripple out to all label selector consumers.

Having a separate spot in the API with key/operator fields that explicitly means "this label selector but with the value pulled from the pod's label" seems better to me than changing semantics of the existing label selector field.

I do wonder if it would make sense to have the API server actually populate the peer label selector field with the pod's label values on create. For example, creating this pod:
```yaml
metadata:
  labels:
    tenant: foo
affinity:
  podAffinity:      # ensures the pods of this tenant land on the same node pool
    requiredDuringSchedulingIgnoredDuringExecution:
      - labelSelector:
          matchExpressions:
          - key: otherlabel
            operator: In
            values: ["othervalue"]
        matchLabelSelectors:
          - key: tenant
            operator: In
```
would result in this pod being persisted:
```diff
 metadata:
   labels:
     tenant: foo 
 affinity:
   podAffinity:      # ensures the pods of this tenant land on the same node pool
     requiredDuringSchedulingIgnoredDuringExecution:
       - labelSelector:
           matchExpressions:
           - key: otherlabel
             operator: In
             values: ["othervalue"]
+          - key: tenant
+            operator: In
+            values: ["foo"]
         matchLabelSelectors:
           - key: tenant
             operator: In
```

That would mean anything reading pods and understanding how they are scheduled wouldn't have to learn about this new field, and the podAffinity / podAntiAffinity fields would continue to contain all the scheduling info

Reply:
+1 on doing this in apiserver.
But then we should do the same in https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3243-respect-pod-topology-spread-after-rolling-upgrades

Reply:
(note that populating/merging with user specified fields on write in the server is usually not a great idea, but could be ok here specifically on pods since the field is immutable after create… also need to consider what would happen with an old client writing pods that didn't know about the new field and maybe copy the old value to the new pod in prepareforupdate)

Reply:
That would simplify the implementation, but wouldn't users be surprised to see that duplicated in the labelSelector when reading back the spec?

Reply:
would need to game it out for sure, but users would also be surprised to see their affinity/antiAffinity rules get ignored, which is what all existing API clients making scheduling decisions on pods would do until updated to understand this new field (alternate schedulers, autoscalers, etc), which might *never* happen

Reply:
Sure.

> But then we should do the same in https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3243-respect-pod-topology-spread-after-rolling-upgrades

It requires people to change the component where they need to pass the feature gate, doesn't it? 
If so, should we do this for PodAffinity only for now, and do this for PodTopologySpread after KEP-3243 is graduated to GA?

Reply:
If "adding a new enum to label selector" has a huge blasting radius, and given we're gonna use API server for field population, is "matchLabelKeys" **_plus_** "notMatchLabelKeys" an alternative? so we don't need to restrict the operator validation on In/NotIn, and also the API field is similar with what we introduced in PodTopologySpread.

```go
matchLabelKeys: []  -> equivalent to a slice of selectors, all with operator In
notMatchLabelKeys: [] -> equivalent to a slice of selectors, all with operator NotIn
```

Reply:
Sorry, missed this comment.
But, I prefer the idea of keeping `matchLabelKeys` + adding `notMatchLabelKeys` to `matchLabelSelectors` because
> the API field is similar with what we introduced in PodTopologySpread.

makes both easier to understand instead of making a difference (matchLabelKeys vs matchLabelSelectors) between them to specify a similar thing.

@alculquicondor @ahg-g What do you think?

Reply:
I'll defer to @ahg-g and @msau42

Reply:
@ahg-g @msau42 kindly ping. 🙏

Reply:
If we can find a more intuitive name for `notMatchLabelKeys`, then I am onboard. Otherwise I would stick with the key+operator approach.

Reply:
- distinctLabelKeys
- ignoreSameValueLabelKeys
- ...? (My English skill can put out only them 😓